appscale/tools/agents/azure_agent.py

```diff
@@ -241,6 +248,106 @@ def configure_instance_security(self, parameters):
     resource_client.providers.register(self.MICROSOFT_COMPUTE_RESOURCE)
     resource_client.providers.register(self.MICROSOFT_NETWORK_RESOURCE)
 
+  def attach_disk(self, parameters, disk_name, instance_id):
+    """ Attaches the persistent disk specified in 'disk_name' to this virtual
+    machine.
+    Args:
+      parameters: A dict containing values necessary to authenticate with the
+        underlying cloud.
+      disk_name: A str naming the managed disk to attach to this machine.
+      instance_id: A str naming the id of the instance that the disk should be
+        attached to.
+    Returns:
+      A str indicating a symlink to where the persistent disk has been
+      attached to.
+    Raises:
+      AgentRuntimeException if the disk cannot be attached due to a
+      CloudError or does not have a successful provisioning state.
+    """
+    credentials = self.open_connection(parameters)
+    resource_group = parameters[self.PARAM_RESOURCE_GROUP]
+    subscription_id = str(parameters[self.PARAM_SUBSCRIBER_ID])
+    compute_client = ComputeManagementClient(credentials, subscription_id)
+    virtual_machine = compute_client.virtual_machines.get(resource_group,
+                                                          instance_id)
+    storage_profile = virtual_machine.storage_profile
+
+    # pick the LUN (this must be unique)
+    existing_luns = set()
+    for disk in storage_profile.data_disks:
+      # While we're looping check if disk is already attached and return the
+      # symlink if it is.
+      if disk.name == disk_name:
+        return '/dev/disk/azure/scsi1/lun{}'.format(disk.lun)
+      existing_luns.add(disk.lun)
+
+    # Get the first number not being used as a LUN.
+    the_chosen_lun = next(ifilterfalse(existing_luns.__contains__, count(1)))
```

Comment:
It took me a while to figure out what this was doing (i've never used ifilterfalse/count). That's pretty cool.

I wonder if the following is a little more clearer and avoids latching into an internal function:
```
the_chosen_lun = next(ifilter(lambda lun: lun not in existing_luns, count(1)))
```
Reading it out loud: get the next element from an iterator where the lun is not in existing_luns, starting from 1.

p.s. nice variable name